File: ahoy-ui/projects/ahoy-components/src/lib/environments/environment.service.ts
/*
 * Copyright  2022 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import {Injectable} from '@angular/core';
import {EMPTY, Observable} from 'rxjs';
import {catchError, map, tap} from 'rxjs/operators';
import {Cluster} from '../clusters/cluster';
import {Notification} from '../notifications/notification';
import {NotificationsService} from '../notifications/notifications.service';
import {RecentReleasesService} from '../release-manage/recent-releases.service';
import {LoggerService} from '../util/logger.service';
import {RestClientService} from '../util/rest-client.service';
import {DeleteOptions, DuplicateOptions, Environment, MoveOptions} from './environment';
import {ErrorUtil} from '../util/error-util';

@Injectable({
  providedIn: 'root'
})
export class EnvironmentService {

  constructor(
    private log: LoggerService,
    private notificationsService: NotificationsService,
    private recentReleasesService: RecentReleasesService,
    private restClient: RestClientService) {
  }

  getAll(): Observable<Environment[]> {
    return this.restClient.get<any>('/data/environments?projection=environmentSummary&sort=orderIndex&sort=id').pipe(
      map(response => response._embedded.environments as Environment[]),
      tap((envs) => this.log.debug(`fetched ${envs.length} environments`))
    );
  }

  getAllForPromotion(releaseId: number): Observable<Environment[]> {
    const url = `/data/environments/search/forPromotion?releaseId=${releaseId}`;
    return this.restClient.get<any>(url).pipe(
      map(response => response._embedded.environments as Environment[]),
      tap((envs) => this.log.debug(`fetched ${envs.length} environments`))
    );
  }

  get(id: number): Observable<Environment> {
    const url = `/data/environments/${id}?projection=environmentSimple`;
    return this.restClient.get<Environment>(url).pipe(
      tap((env) => {
        this.log.debug('fetched environment', env);
      })
    );
  }

  save(environment: Environment): Observable<Environment> {
    if (!environment.id) {
      this.log.debug('saving environment: ', environment);
      return this.restClient.post<Environment>('/data/environments', environment).pipe(
        tap((env) => this.log.debug('saved new environment', env))
      );

    } else {
      this.log.debug('updating environment: ', environment);
      const url = `/data/environments/${environment.id}`;
      return this.restClient.put(url, environment).pipe(
        tap((env) => this.log.debug('updated environment', env))
      );
    }
  }

  delete(environment: Environment, deleteOptions: DeleteOptions): Observable<Environment> {
    this.log.debug('deleting environment: ', environment);

    const id = environment.id;
    const url = `/api/environments/delete/${id}`;

    return this.restClient.delete<Environment>(url, deleteOptions).pipe(
      tap((deletedEnvironment) => {
        this.log.debug('deleted environment', deletedEnvironment);
        this.recentReleasesService.refresh();
      }),
      catchError((error) => {
        if (!ErrorUtil.is500Error(error)) {
          const text = `Failed to delete environment ${environment.key}`;
          this.notificationsService.notification(new Notification(text, error));
        }
        return EMPTY;
      })
    );
  }

  move(environment: Environment, moveOptions: MoveOptions): Observable<Environment> {
    this.log.debug(`moving environment: ${environment.key} to cluster: ${moveOptions.destClusterId}`);

    const url = `/api/environments/${environment.id}/move`;

    return this.restClient.post<Environment>(url, moveOptions, true).pipe(
      tap((destEnvironment) => {
        this.log.debug('moved environment', destEnvironment);
      }),
      catchError((error) => {
        if (!ErrorUtil.is500Error(error)) {
          const text = `Failed to move environment ${environment.key} to cluster`;
          this.notificationsService.notification(new Notification(text, error));
        }
        return EMPTY;
      })
    );
  }

  duplicate(sourceEnvironment: Environment, destEnvironment: Environment, duplicateOptions: DuplicateOptions): Observable<Environment> {
    this.log.debug('duplicating environment: ', sourceEnvironment);

    const url = `/api/environments/duplicate/${sourceEnvironment.id}/${destEnvironment.id}`;

    return this.restClient.post<Environment>(url, duplicateOptions, true).pipe(
      tap((duplicatedEnvironment) => {
        this.log.debug('duplicated environment', duplicatedEnvironment);
        const text = `${duplicatedEnvironment.key} ` + `was duplicated in cluster ${(duplicatedEnvironment.cluster as Cluster).name}`;
        this.notificationsService.notification(new Notification(text));
      }),
      catchError((error) => {
        const text = `Failed to duplicate environment ${destEnvironment.key} from environment: ${sourceEnvironment.key}`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  updateOrderIndex(environment: Environment): Observable<Environment> {
    this.log.debug('updating environment orderIndex: ', environment);

    const url = `/api/environments/${environment.id}/updateOrderIndex?orderIndex=${environment.orderIndex}`;

    return this.restClient.put<Environment>(url, null).pipe(
      tap(() => {
        this.log.debug('updated environment orderIndex', environment);
      }),
      catchError((error) => {
        const text = `Failed to update environment ${environment.key} order`;
        this.notificationsService.notification(new Notification(text, error));
        return EMPTY;
      })
    );
  }

  link(id: number): string {
    return this.restClient.getLink('/data/environments', id);
  }
}

File: ahoy-components/src/main/java/za/co/lsd/ahoy/server/environments/EnvironmentService.java
/*
 * Copyright  2022 LSD Information Technology (Pty) Ltd
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package za.co.lsd.ahoy.server.environments;

import lombok.extern.slf4j.Slf4j;
import org.springframework.data.rest.core.annotation.RepositoryEventHandler;
import org.springframework.data.rest.webmvc.ResourceNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import za.co.lsd.ahoy.server.cluster.Cluster;
import za.co.lsd.ahoy.server.cluster.ClusterRepository;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentRelease;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentReleaseId;
import za.co.lsd.ahoy.server.environmentrelease.EnvironmentReleaseRepository;
import za.co.lsd.ahoy.server.release.DeployOptions;
import za.co.lsd.ahoy.server.release.ReleaseService;
import za.co.lsd.ahoy.server.releases.ReleaseVersion;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RepositoryEventHandler
@Slf4j
public class EnvironmentService {
	private final EnvironmentRepository environmentRepository;
	private final EnvironmentReleaseRepository environmentReleaseRepository;
	private final ClusterRepository clusterRepository;
	private final ReleaseService releaseService;

	public EnvironmentService(EnvironmentRepository environmentRepository, EnvironmentReleaseRepository environmentReleaseRepository, ClusterRepository clusterRepository, ReleaseService releaseService) {
		this.environmentRepository = environmentRepository;
		this.environmentReleaseRepository = environmentReleaseRepository;
		this.clusterRepository = clusterRepository;
		this.releaseService = releaseService;
	}

	@Transactional
	public Environment delete(Long environmentId) {
		Environment environment = environmentRepository.findById(environmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find environment: " + environmentId));

		return delete(environment);
	}

	@Transactional
	public Environment delete(Environment environment) {
		log.info("Deleting environment: {}", environment);

		undeployReleasesFrom(environment);

		environmentRepository.delete(environment);
		return environment;
	}

	@Transactional
	public Environment duplicate(Long sourceEnvironmentId, Long destEnvironmentId, DuplicateOptions duplicateOptions) {
		Environment sourceEnvironment = environmentRepository.findById(sourceEnvironmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find source environment: " + sourceEnvironmentId));

		Environment destEnvironment = environmentRepository.findById(destEnvironmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find destination environment: " + destEnvironmentId));

		log.info("Duplicating environment {} to environment {} with options {}", sourceEnvironment, destEnvironment, duplicateOptions);

		List<EnvironmentRelease> sourceEnvironmentReleases = sourceEnvironment.getEnvironmentReleases();
		for (EnvironmentRelease sourceEnvironmentRelease : sourceEnvironmentReleases) {

			EnvironmentRelease newEnvironmentRelease = new EnvironmentRelease(destEnvironment, sourceEnvironmentRelease.getRelease());
			environmentReleaseRepository.save(newEnvironmentRelease);

			if (duplicateOptions.isCopyEnvironmentConfig()) {
				releaseService.copyEnvironmentConfig(sourceEnvironmentRelease, newEnvironmentRelease);
			}
		}
		return destEnvironment;
	}

	@Transactional
	public Environment move(Long environmentId, MoveOptions moveOptions) {
		Environment environment = environmentRepository.findById(environmentId)
			.orElseThrow(() -> new ResourceNotFoundException("Could not find environment: " + environmentId));

		Cluster cluster = clusterRepository.findById(moveOptions.getDestClusterId())
			.orElseThrow(() -> new ResourceNotFoundException("Could not find destination cluster: " + moveOptions.getDestClusterId()));

		log.info("Moving environment: {} to destination cluster {}", environment, cluster);

		Map<EnvironmentReleaseId, ReleaseVersion> previouslyDeployedReleases = undeployReleasesFrom(environment);

		environment.setCluster(cluster);
		environment = environmentRepository.save(environment);

		if (moveOptions.isRedeployReleases() && !previouslyDeployedReleases.isEmpty()) {
			deployReleasesTo(environment, previouslyDeployedReleases);
		}

		return environment;
	}

	@Transactional
	public void updateOrderIndex(Long environmentId, Double orderIndex) {
		environmentRepository.updateOrderIndex(environmentId, orderIndex);
	}

	private Map<EnvironmentReleaseId, ReleaseVersion> undeployReleasesFrom(Environment environment) {
		Map<EnvironmentReleaseId, ReleaseVersion> previouslyDeployedReleases = new HashMap<>();

		log.info("Checking if there are deployed releases in {}", environment.getKey());
		environment.getEnvironmentReleases().forEach((environmentRelease) -> {
			if (environmentRelease.hasCurrentReleaseVersion()) {
				log.info("Release {} is currently deployed in {}, undeploying...", environmentRelease.getRelease().getName(), environment.getKey());

				previouslyDeployedReleases.put(environmentRelease.getId(), environmentRelease.getCurrentReleaseVersion());

				try {
					releaseService.undeploy(environmentRelease.getId());
				} catch (Exception e) {
					throw new EnvironmentException("Failed to undeploy " + environmentRelease + " from " + environment, e);
				}
			}
		});

		return previouslyDeployedReleases;
	}

	private void deployReleasesTo(Environment environment, Map<EnvironmentReleaseId, ReleaseVersion> previouslyDeployedReleases) {
		environment.getEnvironmentReleases().forEach((environmentRelease) -> {
			if (previouslyDeployedReleases.containsKey(environmentRelease.getId())) {
				ReleaseVersion releaseVersion = previouslyDeployedReleases.get(environmentRelease.getId());

				log.debug("Release {}:{} was previously deployed in {}, redeploying...", environmentRelease.getRelease().getName(), releaseVersion.getVersion(), environment.getKey());

				try {
					String message = String.format("Redeployed %s:%s to %s in %s after moving environment to new cluster",
						environmentRelease.getRelease().getName(),
						releaseVersion.getVersion(),
						environment.getKey(),
						environment.getCluster().getName());

					DeployOptions deployOptions = new DeployOptions(releaseVersion.getId(), message);
					releaseService.deploy(environmentRelease.getId(), deployOptions);
					log.info(message);
				} catch (Exception e) {
					throw new EnvironmentException("Failed to deploy " + environmentRelease + " to " + environment, e);
				}
			}
		});
	}
}
